voltdb/src/main/java/com/yahoo/ycsb/db/voltdb/YCSBSchemaBuilder.java

```diff
@@ -0,0 +1,255 @@
+ * LICENSE file.
+ */
+package com.yahoo.ycsb.db.voltdb;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.util.jar.Attributes;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.voltdb.client.Client;
+import org.voltdb.client.ClientResponse;
+import org.voltdb.client.ProcCallException;
+
+/**
+ * Utility class to build the YCSB schema.
+ * 
+ */
+public final class YCSBSchemaBuilder {
+
+  private static final String PROCEDURE_GET_WAS_NOT_FOUND = "Procedure Get was not found";
+
+  private static final Charset UTF8 = Charset.forName("UTF-8");
+
+  private static YCSBSchemaBuilder instance = null;
+
+  private final String createTableDDL = "CREATE TABLE Store (keyspace VARBINARY(128)  NOT NULL\n"
+      + ",   key      VARCHAR(128)    NOT NULL,   value    VARBINARY(2056) NOT NULL\n"
+      + ",   PRIMARY KEY (key, keyspace));";
+
+  private final String partitionTableDDL = "PARTITION TABLE Store ON COLUMN key;\n";
+
+  private final String createGetDDL = "CREATE PROCEDURE Get PARTITION  ON TABLE Store COLUMN key PARAMETER 1\n"
+      + "AS SELECT value FROM Store WHERE keyspace = ? AND key = ?;";
+
+  private final String createPutDDL = "CREATE PROCEDURE PARTITION ON TABLE Store COLUMN key PARAMETER 1\n"
+      + "FROM CLASS com.yahoo.ycsb.db.voltdb.procs.Put;";
+
+  private final String createScanDDL = "CREATE PROCEDURE PARTITION ON TABLE Store COLUMN key \n"
+      + "FROM CLASS com.yahoo.ycsb.db.voltdb.procs.Scan;";
+
+  private final String createScanAllDDL = "CREATE PROCEDURE \n" + "FROM CLASS com.yahoo.ycsb.db.voltdb.procs.ScanAll;";
+
+  private final String[] ddlStatements = {createTableDDL, partitionTableDDL };
+
+  private final String[] procStatements = {createGetDDL, createPutDDL, createScanDDL, createScanAllDDL };
+
+  private final String[] jarFiles = {"Put.class", "Scan.class", "ScanAll.class", "ByteWrapper.class" };
+
+  private final String jarFileName = "ycsb-procs.jar";
+
+  private Client voltClient;
+
+  private Logger logger = LoggerFactory.getLogger(YCSBSchemaBuilder.class);
+
+  /**
+   * Utility class to build the YCSB schema.
+   * 
+   * @author srmadscience / VoltDB
+   *
+   */
+  private YCSBSchemaBuilder(Client c) {
+    super();
+    this.voltClient = c;
+  }
+
+  /**
+   * We use a single instance of YCSBSchemaBuilder because YCSB can spawn an
+   * arbitrary number of threads that will each try and call init(). Our single
+   * YCSBSchemaBuilder instance remembers requests to load classes and DDL and
+   * only does it once.
+   * 
+   * @param c
+   * @return YCSBSchemaBuilder instance
+   */
+  public static YCSBSchemaBuilder getInstance(Client c) {
```

Comment:
This way of memoizing an instance isn't thread safe.

Does this need to be an actual singleton or are you just trying to get reuse out of the work of making it? If the former then you should use the singleton holder idiom. If the latter then make instance an AtomicRef.

From the ddl load method I guess it needs to be a singleton?

Reply:
Should be fixed by 8c61359